### PR TITLE
Code pieces must be re-written to transition MaxText to Support PyTorch/XLA

### PR DESCRIPTION
**Pull request by miladm**
Added comments on what files and code pieces must be re-written to transition MaxText to Support PyTorch/XLA

### CODE REVIEW

File: MaxText/sequence_packing.py [hunk 1]
```diff
@@ -16,6 +16,9 @@
 
 """Packed Sequence Op."""
 
+# can we use this script out of the box for pytorch? if we want to stay away
+# from tf.data, we should explore further on how to repurpose this helper scrpit
```
Comment (miladm) [inline MaxText/sequence_packing.py:20]: if we were to skip going through tf.data, how much work is repurposing this script for pytorch?

Comment (khatwanimohit) [reply MaxText/sequence_packing.py:20]: This script essentially contains map operations for packing multiple rows into one until the length matches the max_sequence_length, This can be done on huggingface's datasets. Should be 3 days of work

File: MaxText/input_pipeline.py [hunk 2]
```diff
@@ -162,7 +168,7 @@ def filter_fn(x):
   # Return multi-host jax.Array prep iterator
   return multihost_gen
 
-
+# this method needs a rewrite in pytorch land (assuming no use of tf.data)
```
Comment (khatwanimohit) [inline MaxText/input_pipeline.py:171]: we can use huggingface datasets library to get C4: https://huggingface.co/datasets/c4

File: MaxText/input_pipeline.py [hunk 3]
```diff
@@ -84,6 +88,8 @@ def _normalize_features(features):
 # -----------------------------------------------------------------------------
 
 
+# this method needs a rewrite in pytorch land (assuming no use of tf.data and
+# jax api)
```
Comment (khatwanimohit) [inline MaxText/input_pipeline.py:92]: Most of the pre-process operations like map, filter, padding are easily transferrable as they are available for huggingface datasets as well.

File: MaxText/multihost_dataloading.py [hunk 4]
```diff
@@ -20,6 +20,9 @@
 Adapted from Sholto's:
 https://github.com/sholtodouglas/multihost_dataloading
 """
+
+# I expect this script must be rewritten for PyTorch/XLA
```
[the remark this answers is not in the document]
Comment (khatwanimohit) [reply MaxText/multihost_dataloading.py:24]: This script deals with how every host handles its shards of dataset and divides it among its local devices, further creating a global tensor representing a batch with the right sharding_spec. I have recently started this work and might take 1,2 weeks to land ( might be less)

File: MaxText/tokenizer.py [hunk 5]
```diff
@@ -16,6 +16,8 @@
 
 """Provides op for tokenizing a dataset."""
 
+# I expect this script must be rewritten for PyTorch/XLA
+
```
Comment (khatwanimohit) [inline MaxText/tokenizer.py:20]: We can easily re-write this script: https://pytorch.org/text/main/transforms.html#sentencepiecetokenizer. Should be 1,2 days of work.

File: MaxText/configs/base.yml [hunk 6]
```diff
@@ -14,6 +14,9 @@
 
 # This sentinel is a reminder to choose a real run name.
 # If there is already a checkpoint under this run, that checkpoint will auto-resume.
+
+# we need to port this script to the pytorch model
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/configs/base.yml:18]: The logical axis rules dictate how different tensors are sharded (what axis are sharded via DP or FSDP or TP). Sharding is necessary for PyTorch, altho I think it would be okay if it's not configurable and just uses the Maxtext default (which will still take significant work that Mohit is more familiar with)

File: MaxText/layers.py [hunk 7]
```diff
@@ -222,6 +223,8 @@ def __call__(self, inputs: Array) -> Array:
 
     contract_ind = tuple(range(0, len(axis)))
 
+    # quantization - maxtext specific? for short term purposes, bf16 should be
+    # sufficient?
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply MaxText/layers.py:227]: We are currently performing int8 experiments now to see the potential benefits. All perf numbers are based on bf16 - I think its okay if pytorch doesn't support int8 to start, but at some point it should since its likely there is a large performance benefit.

File: MaxText/multihost_dataloading.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (miladm) [reply MaxText/multihost_dataloading.py:24]: do you have an estimate of effort?

File: MaxText/layers.py [hunk 8]
```diff
@@ -350,6 +355,7 @@ def query_init(*args):
     value = checkpoint_name(value, 'value_proj')
 
     if decode:
+      # pytorch needs kv cache
```
Comment (miladm) [inline MaxText/layers.py:358]: @jonb377 wdyt?
